### ISSUE TITLE
Terminating with unexpected exception

### ISSUE THREAD

**Issue by ruslo** [cpp-netlib/cpp-netlib#358]
Example `simple_wget` crashing from time to time with error:
`boost thread: trying joining itself: Resource deadlock avoided`
### How to reproduce

I turn off wi-fi and run command:
 `simple_wget https://example.com/some/archive.tar.gz`

Usually command exits with code 1 and message:

```
Saving to: archive.tar.gz
Host not found (authoritative)
```

which is expected and good, but sometimes application crash in libc++ with message:

```
libc++abi.dylib: terminating with unexpected exception of type boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::thread_resource_error> >: boost thread: trying joining itself: Resource deadlock avoided
Process 75882 stopped
```

Backtrace can be found [here](http://pastebin.com/nGhy61mu)

I can guess that exception which is thrown from `join` trigger `std::terminate`
because destructor has _nothrow_ type (async_impl.h):

``` cpp
      ~async_client() throw ()
      {
        sentinel_.reset();
        if (lifetime_thread_.get()) {
          lifetime_thread_->join(); // BOOM!
          lifetime_thread_.reset();
        }
      }
```
### Some tests
- 5 crash from 1000 run
- 36 crash from 10000 run
### Environment
- Mac OS X 10.9.1, boost 1.55 (debug build with libc++), openssl (system)
- iOS, boost 1.55 (debug build with libc++), openssl 1.0.1e
- I'm using `cpp-netlib-0.10.1`, but `cpp-netlib-0.11.0-rc3` has same behaviour

Also I've tried it on my virtual machine with Mac OS X 10.8.5 (which not using wi-fi),
there is **no** crashes.

**Comment by olegabr** [comment]
I've got the same error in the 0.13-release branch now.

```
terminate called after throwing an instance of 'std::system_error'
  what():  Resource deadlock avoided
```

The patch that fixes it that was applied in the 0.11-devel here: https://github.com/cpp-netlib/cpp-netlib/commit/1b11ec63aaed2b8acdcdb3451f06ca81e7e02789#diff-c272e324d5e4a6288abe1e22553c4027
was reverted without any explanations here: https://github.com/cpp-netlib/cpp-netlib/commit/2be7e300be52f6d0c347a37e567b52889bfab4f7#diff-c272e324d5e4a6288abe1e22553c4027

The question is why? And should we apply it back?
